### ISSUE TITLE
I think I know how making 12 space lookup tables can be made easier.

### ISSUE THREAD

**Issue by 1h0si** [macroxue/2048-ai#8]
I think if you split the lookup tables into each individual layer by the sum of tiles, you will need much less memory to create the large tables. This could allow 12 space lookup tables with two free rows and columns, which would allow for some much better strategies to appear

**Comment by macroxue** [comment]
Thanks for thinking about it. I'm not sure if I understand the proposal though. Can you please be more specific? Diagrams showing individual layers with calculations on the memory usage will definitely help.

**Comment by 1h0si** [comment]
If you make these lookup tables the same way I would think they're made (generating positions in ascending order then solving them descending), then you would only need to consider three layers at any given time. 

When you're generating positions, for example, you might be generating positions from a sum of 4. You'd generate all of the 2 spawns and all of the 4 spawns, which are on layers 6 and 8 respectively. You'll never need to consider layer 2 again, and the same applies anywhere else. When solving a position, for example, on layer 386, you only need to consider 2 spawns leading to 388 and 4 spawns leading to 390. No other layers besides those three need to be considered at this time.

To show the difference I've made LL12-512 and LL12-1024 Tables in the past, using a tool that only had three layers in memory at a time, and the difference was much greater. while the tables were about 300 and 900 GB respectively, the program never used more than 10 GB of memory. Unfortunately, this program is also significantly slower (DPDF10 taking 2 hours versus your 25 minutes). I'm unsure whether there's any correlation between the two. LL12-1024 has approximately 55 billion positions in the tree

I hope this made it clearer

**Comment by macroxue** [comment]
I think I get the idea, which is a breath-first traversal and friendly to swapping.

Since you were able to move two rows and two columns, did you manage to get higher scores?

**Comment by 1h0si** [comment]
I haven't been able to see exactly how helpful this is in real games, but LL12-512 outperforms DPDF10-128 (0.999975455 vs 0.99989) and LL12-1024 outperforms DPDF10-256 (0.99519 vs 0.9938). DPDF10-512 surprisingly outperforms LL12-2048 (0.8337 vs 0.829), although if you are forcing strict DPDF/gap with 12 spaces, DPDF would probably have a success rate closer to 0.821, lower than LL-2048. We currently do not know how well LL-4096 performs, but using endgames alone, we have a lower bound of about 3.6% for the 65536 tile

**Comment by macroxue** [comment]
3.6% sounds about right. The lower bound seems pretty tight already. I was a bit lucky to get 3.5% with 1000 runs of DPDF, which gives me less motivation to shoot for 3.6% or maybe 3.7%. :)

**Comment by macroxue** [comment]
I read from multiple sources that LL-4096 with T-formation reaches the 65536 tile with nearly 2X success rate (~6%). Did a quick calculation on the number of positions and took a shortcut to stop the calculation when the sum of small tiles is 1.5X of the target tile, which didn't seem to hurt the success rate of the initial position.

| Target   | # Positions |
|-------------|-------------|
| 128   | 775,797,081 |
| 256   | 4,342,925,560 |
| 512   | 17,427,629,016 |
| 1024 | 54,333,786,618 |
| 2048 | 140,864,964,607 |
| 4096 | 317,980,987,690 |

For 1024, 54.3 billion is very close to 55 billion you mentioned earlier, so I assume the calculation is correct. For 4096, even if a position takes only 8 bytes, 318 billion positions will need 2.544 TB of storage and 2 days by my estimate. With this much requirement, it becomes much less practical for most people. I'm trying to see if there is a smarter way.

**Comment by 1h0si** [comment]
This is true. But better formations have since been found. A truly perfect ai is estimated to reach 65536 with a success rate between 8.5% and 9%. The lower estimate was reached through a multitude of tables. The largest factor here was a formation resembling
```
s s s s
s s s s
L s s L
L s s L
```

Allowing one column to shift from here yields a success rate of nearly 14% for the final 4096 (compared to 11.97% for a square in the corner), and 16.665% for the final 2048. 

Other large lookup tables have also been made, including tables that allow large moving tiles. Using 11 small and 5 arbitrarily large tiles, all moving (free11), we were able to reach much higher success rates than ever previously thought

| endgame (final n/32) | optimal start | Snake start | DPDF start |
| -------- | ------- | -------- | ------- |
| 8192  |  0.99999842  |  0.99999842 | 0.99997733 |
| 16384 | 0.99833  |  0.99769 | 0.99713 |
| 32768    | 0.879205 | 0.874621 | 0.863015 |

The upper bound was reached by multiplying all appearances of these endgames together. In addition, we used the smaller free10 (maximum success rate of 0.2007) to finish the 65536 endgame (with another appearance of a 32768 free11 endgame for the penultimate 1024). Combining all of these together yields an upper bound of about 0.090275. Considering that a perfect AI can quite reliably rearrange and recover with five large tiles (as shown by the DPDF starts with significantly higher success rates than they would be without rearranging), it must be almost guaranteed for a perfect AI to reach the formation shown above. Chances are, a perfect AI is between 8.5% and 9% due to this likely scenario

Re: generating large tables

Your numbers appear to be roughly correct. However, the fastest table generating tools are faster than this, generating LL-4096 in about 12 hours. However, you can actually get away with stopping generating much sooner than you think. The success rate loss is virtually nonexistent if you generate to a sum just a constant 96 higher than your target tile.

**Comment by macroxue** [comment]
Thanks for sharing the amazing progress! I had never guessed that 9% success rate for 65536 was possible. You guys are really pushing the limit. Maybe one day the 4x4 board will be completely solved.

The symmetric T-formation makes a lot of sense after I played with the asymmetric T-formation with LL-4096. Since free10 and free11 are so good, I guess someone is going to work on free12 next?
